Improved breakpoint gutter for CodeEdit.

EricEzaM opened the pull request:
- Added preview of breakpoint placement on hover. 
- Added override of icon for the Editor to use an SVG so it is no longer blurry.

https://user-images.githubusercontent.com/41730826/134858617-fb94a0af-8a4d-4b83-9521-cffe712a9d69.mp4

*Bugsquad edit: This closes https://github.com/godotengine/godot-proposals/issues/533.*

bruvzg commented on editor/icons/Breakpoint.svg:
```diff
@@ -0,0 +1,16 @@
+<svg
+	enable-background="new 0 0 16 16"
+	height="16"
+	viewBox="0 0 16 16"
+	width="16"
+	xmlns="http://www.w3.org/2000/svg"
+>
+	<path
+		d="m15 8c0 3.866-3.134 7-7 7s-7-3.134-7-7 3.134-7 7-7 7 3.134 7 7"
+		fill="#e1e1e1"
+	/>
+	<path
+		d="m12 8c0 2.209-1.791 4-4 4s-4-1.791-4-4 1.791-4 4-4 4 1.791 4 4"
+		fill="#f6f6f6"
+	/>
+</svg>
```
`svgo` can cut it to a half size, 0.34 KiB → 0.186 KiB

```suggestion
<svg height="16" width="16" xmlns="http://www.w3.org/2000/svg"><path d="M15 8A7 7 0 1 1 1 8a7 7 0 0 1 14 0" fill="#e1e1e1"/><path d="M12 8a4 4 0 1 1-8 0 4 4 0 0 1 8 0" fill="#f6f6f6"/></svg>

```

EricEzaM replied:
Oh sorry, I auto-formatted it in VS code to easily view it but forgot to compress it back, thanks.

Paulb23 reviewed (changes requested):
Can't speak for adding a new icon. But we should probably not draw the hover icon if the user is selecting / scrolling, can call `is_dragging_cursor` to check. 

Also it going to be a lot of updates, not sure if we should limit theses in some way.

EricEzaM commented:
@Paulb23 I put a breakpoint on the `TextEdit` `NOTIFICATION_DRAW` logic and it appears to get called every frame anyway. I added the dragging logic as you describe.

Paulb23 reviewed (commented):
> NOTIFICATION_DRAW logic and it appears to get called every frame anyway

Make sure to disable caret blink, otherwise it will re-draw every `0.5` secs.

EricEzaM committed:
Improved breakpoint gutter for CodeEdit.

Added preview of breakpoint placement on hover. Added override of icon for the Editor to use an SVG so it is no longer blurry.

EricEzaM commented:
Ha! I just realised it _wasn't_ drawing every frame. When focused Godot after setting the breakpoint, the cursor would draw, hitting the breakpoint. When the breakpoint was hit, Visual Studio would take focus. Then when "continuing" Godot would trigger the event which said it lost focus, and thus would stop drawing the cursor. Repeat this cycle and it _looked_ like it was every frame, but it wasn't. Oops.

Anyway, I cache the last hovered gutter row into a class instance variable, and only call update when the hovered gutter row changes. Thus it looks the same, but only draws when it needs to change the position of the breakpoint preview, and not on every mouse movement.

Paulb23 commented:
Ah! Yeah that's always fun, race to `alt-tab`  before the debugger continues ;)

We should probably check gutter index as well, otherwise looks good.

Paulb23 commented on scene/gui/text_edit.cpp:
```diff
@@ -1636,6 +1636,30 @@ void TextEdit::gui_input(const Ref<InputEvent> &p_gui_input) {
 		}
+
+		// Check if user is hovering a different gutter, and update if yes.
+		int hovered_row = get_line_column_at_pos(mpos).y;
+		Vector2i current_hovered_gutter = Vector2i(-1, -1);
+
+		int left_margin = style_normal->get_margin(SIDE_LEFT);
+		for (int i = 0; i < gutters.size(); i++) {
+			if (!gutters[i].draw || gutters[i].width <= 0) {
+				continue;
+			}
+
+			if (mpos.x > left_margin && mpos.x <= (left_margin + gutters[i].width) - 3) {
+				// We are in this gutter i's horizontal area.
+				current_hovered_gutter = Vector2i(i, hovered_row);
+				break;
+			}
+
+			left_margin += gutters[i].width;
+		}
```
Sorry, one more nitpick :) but we can skip running most of this if the mouse is not over the gutter.
```suggestion
		Vector2i current_hovered_gutter = Vector2i(-1, -1);
		
		int left_margin = style_normal->get_margin(SIDE_LEFT);
		if (mpos.x <= left_margin + gutters_width + gutter_padding) {
			int hovered_row = get_line_column_at_pos(mpos).y;
			for (int i = 0; i < gutters.size(); i++) {
				if (!gutters[i].draw || gutters[i].width <= 0) {
					continue;
				}

				if (mpos.x > left_margin && mpos.x <= (left_margin + gutters[i].width) - 3) {
					// We are in this gutter i's horizontal area.
					current_hovered_gutter = Vector2i(i, hovered_row);
					break;
				}

				left_margin += gutters[i].width;
			}
		}
```

akien-mga commented:
Thanks!
